src/rgw/rgw_notify.cc

```diff
@@ -2,17 +2,530 @@
+      return -EINVAL;
+    }
+
+    return 0;
+  }
+
+  // set m1 to be the minimum between m1 and m2
+  static int set_min_marker(std::string& m1, const std::string m2) {
+    cls_queue_marker mr1;
+    cls_queue_marker mr2;
+    if (mr1.from_str(m1.c_str()) < 0 || mr2.from_str(m2.c_str()) < 0) {
+      return -EINVAL;
+    }
+    if (mr2.gen <= mr1.gen && mr2.offset < mr1.offset) {
+      m1 = m2;
+    }
+    return 0;
+  }
+
+   class tokens_waiter {
+     const boost::posix_time::time_duration infinite_duration = boost::posix_time::hours(1000);
+     size_t pending_tokens = 0;
+     boost::asio::deadline_timer timer;
+ 
+     struct token {
+       tokens_waiter& waiter;
+       token(tokens_waiter& _waiter) : waiter(_waiter){
+         ++waiter.pending_tokens;
+       }   
+  
+       ~token() {
+         --waiter.pending_tokens;
+         if (waiter.pending_tokens == 0) {
+           waiter.timer.cancel();
+         }   
+       }   
+     };
+  
+   public:
+
+     tokens_waiter(boost::asio::io_context& io_context) : timer(io_context) {}  
+ 
+     void async_wait(spawn::yield_context yield) { 
+       timer.expires_from_now(infinite_duration);
+       boost::system::error_code ec; 
+       timer.async_wait(yield[ec]);
+       // TODO: make sure that the timer was cancelled and not rexpired
+     }   
+ 
+     token make_token() {    
+       return token(*this);
+     }   
+  };
+
+
+  // processing of a specific queue
+  // TODO: use string_view for queue_name
+  void process_queue(spawn::yield_context yield, std::string queue_name) {
+    const auto max_elements = 1024;
+    const std::string start_marker;
+    bool truncated = false;
+    std::string end_marker;
+    std::vector<cls_queue_entry> entries;
+    auto is_idle = true;
+    auto total_entries = 0U;
+    auto pending_work = 0U;
+    tokens_waiter waiter(io_context);
+
+    const auto ret = cls_2pc_queue_list_entries(rados_ioctx, queue_name, start_marker, max_elements, entries, &truncated, end_marker);
+    if (ret < 0) {
+      ldout(cct, 5) << "WARNING: failed to get list of entries in queue: " 
+        << queue_name << ". error: " << ret << " (will retry)" << dendl;
+      goto sched_next;
+    }
+    total_entries = entries.size();
+    ldout(cct, 20) << "INFO: found: " << total_entries << " entries from: " << queue_name << dendl;
+    if (total_entries == 0) {
+      // nothing in the queue
+      goto sched_next;
+    }
+    
+    is_idle = false;
+    {
+      auto has_error = false;
+      auto remove_entries = false;
+      auto entry_idx = 0U;
+      for (auto& entry : entries) {
+        if (has_error) {
+          // bail out on first error
+          break;
+        }
+        // TODO pass entry pointer instead of by-value
+        spawn::spawn(io_context, [this, entry_idx, total_entries, &end_marker, &remove_entries, &has_error, &waiter, entry](spawn::yield_context yield) {
```

Comment:
ok. currently i use a single thread in the pool, but even in case of multiple threads, processing of entries of a for a specific topic must be sequential